### PR TITLE
Fix #10050: Use lowercase pattern variable syntax for quoted patterns

### LINKED ISSUE #10050: Improve syntax of type holes in quoted patterns
## Current state

After #10045 we have the following 2 syntaxis to express a type hole in quote pattern.
The first is written using a `$` on the type.
```Scala
case '{ $x : $T } => '{ val y: T = .. .}
```
`T` and a `given Type[T]` are made available on the right-hand side of the pattern.

The second is provides a workaround to some limitations of the first syntax
```scala
case '{ type $T; $x : `$T` } => '{ val y: T = .. .}
```
Here the `T` is defined as a `type` in the pattern itself.
It allows us to write explicit bounds on such as 
```scala
case '{ type $T <: Seq[Int]; ... } =>
```
It also allows us to have a spliced type that is used in several places 
```scala
case '{ type $T; val x: Map[`$T`, `$T`]; ... } =>
```

On the other side we have known types from outside that are used in patterns which just name the type
```scala
def f[T: Type] = 
   ... match
     case '{ $x : T } => '{ val y: T = .. .}
```

Patterns are composable in the sense that a type extracted from one pattern can be used as a known type of a nested pattern.
```scala
match
  case '{ $x: $T } => 
    ... match 
      case '{ ... : T } => // T extracted from first match used in second
```
## Goals

* Remove the second syntax variant
  * Support bounds in normal holes
  * Support multiple uses of the same splice in normal holes
* Break fake symmetry between `$x` and `$T`
* Keep support of known types as it is
* Type from a pattern should compose with other patterns

### PR DESCRIPTION
**Pull request by nicolasstucki**
Use lower cases for quoted type patterns
```scala
case '{ $x: T } => '{ val y: T = .. .} // T from outside
case '{ $x: t } => '{ val y: t = .. .} // t form within the quotes
case '{ type t; $x: Map[`t`, `t`] } => '{ val y: t = .. .} // t form within the quotes
case '{ $x: `t` } => '{ val y: t = .. .} // t from outside
```

Based on https://github.com/lampepfl/dotty/issues/10050#issuecomment-712885811

### CODE REVIEW

**Comment by smarter** [comment]
> case '{ type t; $x: Map[`t`, `t`] } => '{ val y: t = .. .} // t form within the quotes

Would the following be possible?

```scala
case '{ $x: Map[t, t] } => '{ val y: t = .. .} // t form within the quotes
```
Semantically this makes sense I think: we accept a pattern in all these positions, it's just that it has to be the same one. Normal pattern syntax does not currently allow referring to the same pattern variable multiple times in a pattern, but I recall @LPTK had some ideas on how this could be useful (e.g., when using alternatives).

**Comment by nicolasstucki** [comment]
@smarter in theory it is possible. It is the next step that I want to try. The change needed to support this is unrelated to the change in syntax and I would add it in another PR.

**Comment by LPTK** [comment]
It would be great if the change made to allow repeating type pattern variables would be extended to all patterns, not just quoted patterns.

As for supporting term pattern variables in pattern alternatives, that's [a whole other beast](https://github.com/LPTK/dotty/pull/1)!

File: compiler/src/dotty/tools/dotc/core/NameKinds.scala [hunk 1]
```diff
@@ -320,7 +320,7 @@ object NameKinds {
   val PatMatStdBinderName: UniqueNameKind    = new UniqueNameKind("x")
   val PatMatAltsName: UniqueNameKind         = new UniqueNameKind("matchAlts")
   val PatMatResultName: UniqueNameKind       = new UniqueNameKind("matchResult")
-  val PatMatVarName: UniqueNameKind          = new UniqueNameKind("ev$")
+  val PatMatVarName: UniqueNameKind          = new UniqueNameKind("$given")
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/NameKinds.scala:323]: If this is only used for givens, maybe this should be PatMatGivenName?

File: compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala [hunk 2]
```diff
@@ -185,6 +171,21 @@ trait QuotesAndSplices {
       tree1
   }
 
+  def typedQuotedTypeVar(tree: untpd.Ident, pt: Type)(using Context): Tree =
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala:174]: A documentation comment for this method would be nice.

**Review by smarter — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/NameKinds.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/NameKinds.scala:323]: Technically it is only used to to be able to identify a synthetic name as a pattern variable name.

File: compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala:174]: I will add it

File: compiler/src/dotty/tools/dotc/core/NameKinds.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/NameKinds.scala:323]: But it is true that if it is synthetic, the only way to access it is if it is given. I will rename it.

**Comment by nicolasstucki** [comment]
Rebased and resolved conflicts

**Pull request merged**